File: scripts/_codex-frontier-failures.mjs
import { DatabaseSync } from "node:sqlite";

const file = process.argv[2];
const leadId = process.argv[3];
const pattern = process.argv[4];
if (!file || !leadId) throw new Error("frontier sqlite path and lead id required");

const db = new DatabaseSync(file, { readOnly: true });
const run = db
  .prepare("SELECT id FROM CrawlRun WHERE leadId = ? ORDER BY startedAt DESC LIMIT 1")
  .get(leadId);
const nodes = run && pattern
  ? db
      .prepare(
        `SELECT canonicalUrl, parentUrl, state, relevance, resourceType, httpStatus, retryCount, lastError
           FROM CrawlFrontierNode
          WHERE crawlRunId = ? AND canonicalUrl LIKE ?
          ORDER BY canonicalUrl
          LIMIT 100`
      )
      .all(run.id, `%${pattern}%`)
  : run
  ? db
      .prepare(
        `SELECT canonicalUrl, parentUrl, state, relevance, resourceType, httpStatus, retryCount, lastError
           FROM CrawlFrontierNode
          WHERE crawlRunId = ?
            AND state IN ('TECHNICAL_BLOCKED', 'RETRY_PENDING')
          ORDER BY relevance DESC, state, canonicalUrl
          LIMIT 100`
      )
      .all(run.id)
  : [];
console.log(JSON.stringify(nodes, null, 2));
db.close();

File: scripts/_codex-frontier-failures.py
#!/usr/bin/env python3
import json
import sqlite3
import sys


database_path, lead_id = sys.argv[1:3]
connection = sqlite3.connect(f"file:{database_path}?mode=ro", uri=True)
connection.row_factory = sqlite3.Row
rows = connection.execute(
    """
    SELECT node.crawlRunId, node.canonicalUrl, node.parentUrl,
           node.discoverySource, node.resourceType, node.relevance,
           node.state, node.retryCount, node.lastError,
           node.exclusionReason, node.httpStatus
      FROM CrawlFrontierNode node
      JOIN CrawlRun run ON run.id = node.crawlRunId
     WHERE run.leadId = ?
       AND node.state IN ('TECHNICAL_BLOCKED', 'RETRY_PENDING', 'FETCHING', 'FETCHED')
     ORDER BY node.state, node.relevance, node.canonicalUrl
    """,
    (lead_id,),
).fetchall()
print(json.dumps([dict(row) for row in rows], ensure_ascii=False, indent=2))
connection.close()
